src/iterator.cr

```diff
@@ -1290,4 +1290,91 @@ module Iterator(T)
       self
     end
   end
+
+  # Returns an iterator for each chunked elements where the ends
+  # of chunks are defined by the block, when the block's value
+  # is _truthy_. Thus, the enumerable is sliced just **after** each
+  # time the block returns a _truthy_ value.
+  #
+  # For example, to get chunks that end at each uppercase letter:
+  #
+  # ```
+  # ary = ['a', 'b', 'C', 'd', 'E', 'F', 'g', 'h']
+  # iter = ary.slice_after(&.uppercase?)
+  # iter.next # => ['a', 'b', 'C']
+  # iter.next # => ['d', 'E']
+  # iter.next # => ['F']
+  # iter.next # => ['g', 'h']
+  # iter.next # => Iterator::Stop
+  # ```
+  #
+  # By default, a new array is created and yielded for each slice when invoking `next`.
+  # * If *reuse* is `false`, the method will create a new array for each chunk
+  # * If *reuse* is `true`, the method will create a new array and reuse it.
+  # * If *reuse* is an `Array`, that array will be reused
+  #
+  # This can be used to prevent many memory allocations when each slice of
+  # interest is to be used in a read-only fashion.
+  def slice_after(reuse : Bool | Array(T) = false, &block : T -> B) forall B
+    SliceAfter(typeof(self), T, B).new(self, block, reuse)
+  end
+
+  # :nodoc:
+  class SliceAfter(I, T, B)
+    include Iterator(Array(T))
+
+    def initialize(@iterator : I, @block : T -> B, reuse)
+      @end = false
+      @clear_on_next = false
+
+      if reuse
+        if reuse.is_a?(Array)
+          @values = reuse
```

Comment:
Shouldn't `@clear_on_next = true` here?

Reply:
Hmm... you are supposed to pass an empty array as a reuse value. It's not documented what happens if you pass an array that has some values already, but in this case they will be part of the first chunk. Maybe that's good. I wouldn't worry about this now.

Reply:
The obvious usecase for passing an array as `reuse` is when you've just used the array from another iterator, meaning the array might not be empty. I don't think it's good to pass the burden of calling clear onto the caller, since it's easy to forget here. Also, `clear` on an already-empty array is very cheap, so why not?

Reply:
What if you want to get chunks but want the first chunk to have some elements beforehand? If the iterator calls `clear` there's no way you can do that.

I think we are just inventing use cases. Whenever I used `reuse` I passed a boolean. We can maybe remove the ability to pass an array altogether.

Reply:
If you want the first chunk to have extra elements:
1. that's pretty inflexible, since you can only do it with the first chunk, not the last
2. using a named arg called `reuse: ` is not the most clear way to achieve this.

For me, `reuse` is for avoiding memory allocations only.